### PR TITLE
wait for metaserer ready(loop)

### PR DESCRIPTION
**Pull request by andychow01**
MetaClient instance wait meta serer ready(loop), then do other things like start threadpool.....

### CODE REVIEW

File: src/meta/client/MetaClient.cpp [hunk 1]
```diff
@@ -39,8 +39,24 @@ MetaClient::~MetaClient() {
...
 
 
-void MetaClient::init() {
+void MetaClient::waitForMetadReady() {
     loadDataThreadFunc();
+
+    if (!ready_ && sendHeartBeat_) {
+        while (true) {
+            if (listener_ == nullptr) {
+                VLOG(1) << "Can't send heartbeat due to listener_ is nullptr!";
```
Comment (wadeliuyi) [inline src/meta/client/MetaClient.cpp]: Function “heartbeat” call “auto localHost = listener_->getLocalHost();”, so need continue here

File: src/meta/client/MetaClient.cpp [hunk 2]
```diff
@@ -39,8 +39,24 @@ MetaClient::~MetaClient() {
...
+            }
+            auto ret = heartbeat().get();
+            if (!ret.ok()) {
+                LOG(ERROR) << "Heartbeat failed, status:" << ret.status();
+                ::sleep(1);
```
Comment (wadeliuyi) [inline src/meta/client/MetaClient.cpp]: maybe can use std::this_thread::sleep_for to sleep short

File: src/meta/client/MetaClient.cpp (hunk 1 above)
Comment (andychow01) [reply src/meta/client/MetaClient.cpp]: fixed! Thanks! But I think exit(EXIT_FAILURE) will be more reasonable!

File: src/meta/client/MetaClient.cpp (hunk 2 above)
Comment (andychow01) [reply src/meta/client/MetaClient.cpp]: fixed! Thanks!

File: src/meta/client/MetaClient.cpp [hunk 3]
```diff
@@ -39,8 +39,25 @@ MetaClient::~MetaClient() {
 }
 
 
-void MetaClient::init() {
+void MetaClient::waitForMetadReady() {
     loadDataThreadFunc();
```
Comment (dangleptr) [inline src/meta/client/MetaClient.cpp]: Let's figure out what's ready for Meta client. 
Currently,  meta client sends HB to meta server on storage layer,  but it does not on query engine side. 
On storage side,  meta client ready means that server accept the HB and local cache is filled. 
On graph side,  meta client ready means that local cache is filled. 

So we should handle both situations above.

File: src/meta/client/MetaClient.cpp (hunk 1 above)
Comment (wadeliuyi) [reply src/meta/client/MetaClient.cpp]: > fixed! Thanks! But I think exit(EXIT_FAILURE) will be more reasonable!

so that mean metaclient is strong depend listener?

File: src/meta/client/MetaClient.cpp (hunk 3 above)
Comment (andychow01) [reply src/meta/client/MetaClient.cpp]: Modified! Please review!

File: src/meta/client/MetaClient.cpp [hunk 4]
```diff
@@ -39,8 +41,19 @@ MetaClient::~MetaClient() {
...
+    while (!ready_) {
+        if (sendHeartBeat_) {
+            auto ret = heartbeat().get();
+            if (!ret.ok()) {
+                LOG(ERROR) << "Heartbeat failed, status:" << ret.status();
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                continue;
+            }
+        }  // end if
+        loadDataThreadFunc();
```
Comment (critical27) [inline src/meta/client/MetaClient.cpp]: Maybe you need to sleep a while here? If it a MetaClient of graph, and fail to load data, it will try to call listSpaces all the time?

Comment (andychow01) [reply src/meta/client/MetaClient.cpp]: Now I have added a sleep. Thanks!

File: src/meta/client/MetaClient.cpp (hunk 1 above)
Comment (andychow01) [reply src/meta/client/MetaClient.cpp]: Now, the logic has changed. Metaclient don't strong depend on listener!

File: src/meta/client/MetaClient.cpp [hunk 5]
```diff
@@ -39,8 +41,22 @@ MetaClient::~MetaClient() {
 }
 
 
-void MetaClient::init() {
-    loadDataThreadFunc();
+bool MetaClient::waitForMetadReady() {
+    while (!ready_) {
```
Comment (wadeliuyi) [inline src/meta/client/MetaClient.cpp]: is there any method to stop this loop if the ready_ alway is false?

File: src/meta/test/MetaClientTest.cpp [hunk 6]
```diff
@@ -469,7 +475,7 @@ TEST(MetaClientTest, HeartbeatTest) {
         }
     }
     sleep(FLAGS_heartbeat_interval_secs + 1);
-    ASSERT_EQ(1, ActiveHostsMan::instance()->getActiveHosts().size());
```
Comment (dangleptr) [inline src/meta/test/MetaClientTest.cpp]: Why the changes?
